### ISSUE TITLE
How to define a new palette in TDS5/ncWMS2? (v5.0.0-alpha3)

### ISSUE THREAD

**Issue by danwild** [Unidata/tds#173]
I'm using [v5.0.0-alpha3](https://github.com/Unidata/thredds/releases/tag/v5.0.0-alpha3)

I have a  (very simple) custom colour palette that I was using successfully in TDS4/ncWMS1, however I can't figure out how to do so with TDS5/ncWMS2.

Previously in TDS4, I simply defined my new palette by adding my `.pal` file to `WEB-INF/palettes` and restarting.

Now, there are a couple of things that are confusing in TDS5:

* Despite the ~8 palettes included in `WEB-INF/palettes`, only `default` appears in `GetCapabilities` request (there are a few *styles* but in the form `default-scalar/default`, `colored_contours/default`, etc.).
* A `GetMetadata` request returns 94 palette styles.
* A `GetMap` request using any of the above mentioned 94 styles returns successfully despite not being declared in `GetCapabilities` (although perhaps this is a feature; given that styles take the form type/palette the number of combinations if they were to be listed explicitly might be a little wordy..).
* Some of the palettes defined in `WEB-INF/palettes` are not present in the list returned by `GetMetadata`, so it seems the `WEB-INF/palettes` may not actually be used at all?

I have tried setting alternate `paletteLocationDir`  in `threddsConfig.xml` but seems to have no effect.

I'm trying to find where some of these new palettes are coming from (e.g. `psu-magma`) but had no luck yet.

Any tips much appreciated! 
Thanks,
Dan

**Comment by billyz313** [comment]
Any help on this will be greatly appreciated.  We have upgraded our thredds servers to 5.0, but are currently in fear that we have to downgrade back to 4.6 due to the palettes not being available. @haileyajohnson Can you offer any help please?

**Comment by lesserwhirls** [comment]
This should be supported. I will be able to fully respond this afternoon.

**Comment by lesserwhirls** [comment]
Full configuration of the wms service with TDS version 5 (which uses edal-java) is pretty recent. From what I can tell, the format of the palette files changed between ncWMS and edal-java. The new palette files that ship with the TDS can be found in the edal-java repo at:

https://github.com/Reading-eScience-Centre/edal-java/tree/master/graphics/src/main/resources/palettes

The new palettes look like this:

https://reading-escience-centre.gitbooks.io/ncwms-user-guide/content/04-usage.html#getmap

The palette files in:

https://github.com/Unidata/tds/tree/main/tds/src/main/webapp/WEB-INF/palettes

need to be removed as they are no longer valid, and the code for the TDS will not look in there for palette files.

The new palette format is described in the ncwms2 docs:

https://reading-escience-centre.gitbooks.io/ncwms-user-guide/content/06-development.html#palettes

However, keep in mind those are ncwms2 docs, so not everything you read there will apply to the TDS (for example, the locations to place custom palette files is different).

In order to use custom palette files, you will need to store them in a location on disk that is readable by the user running the TDS, and edit `threddsConfig.xml` to point to that location, something along the lines of:

~~~xml
<WMS>
  ...
  <paletteLocationDir>/my/custom/palettes</paletteLocationDir>
</WMS>
~~~

The TDS documentation is undergoing a lot of changes, and some of these details are part of the documentation for the next release. See:

https://docs.unidata.ucar.edu/tds/5.1/adminguide/adding_wms.html

and

https://docs.unidata.ucar.edu/tds/5.1/adminguide/customizing_wms.html

The location of these docs are subject to change as things continue to move around.

**Comment by billyz313** [comment]
So all 46 palettes should load by default?  In my wms capabilities with or without custom palettes i only get the following three palettes:

- default-scalar/default
- colored_contours/default
- raster/default

Is there something i'm missing to enable them?  

PS.  I have already tried the hex values for the custom palettes and still get nothing,  I assume this is due to the above mentioned and once we get this sorted it will pick up all of them.

**Comment by billyz313** [comment]
@lesserwhirls I notice u point to the 5.1 documentation, is there a place to download the 5.1 war file from?  Is it possible the palettes were not added in for the thredds##5.0.war (which is what we have installed)?

**Comment by haileyajohnson** [comment]
@billyz313 there is no TDS 5.1 yet, 5.0 is the current version. The documentation says 5.1 just because we're in the middle of changing our documentation, and the 5.0 docs don't reflect all the documentation changes yet.

**Comment by billyz313** [comment]
@haileyajohnson , thank you.  I still need assistance getting more than the 3 palettes mentioned above to show up and be usable.  Are there any hints?  Will I have to just wait for 5.1 to come out to be able to use the palettes?

**Comment by haileyajohnson** [comment]
You shouldn't need to wait for a new release - my last comment was a little unclear, but what I was trying to say is that everything posted on this thread, including the linked documentation, is referring to 5.0 (the 5.1 in the url is an unrelated documentation issue). I'm not sure why palettes wouldn't be working for you, maybe @lesserwhirls has some ideas?

**Comment by billyz313** [comment]
Thank you, any assistance would be great, palettes are the only thing holding us from upgrading to 5.0 from 4.6.  If I can just get the 46 default ones that are shipped with the TDS to work it would be a huge leap forward.

**Comment by lesserwhirls** [comment]
Would be able to post the response to `getCapabilities`?

**Comment by lesserwhirls** [comment]
BTW - I am able to add custom palettes to a default TDS installation. I'm not sure why at the moment, but the wms config option for setting the custom palette location 

~~~xml
<WMS>
  ...
  <paletteLocationDir>/my/custom/palettes</paletteLocationDir>
</WMS>
~~~

**does not** work. The default custom palette location **does** work, and is located at `${tds.content.root.path}/thredds/wmsPalettes`.

It also might be helpful to know that there is a mapping between some ncWMS 1.x palette names and edal-java palette names. See https://docs.unidata.ucar.edu/tds/5.1/adminguide/release_notes.html#ncwms

**Comment by billyz313** [comment]
Here is the capabilities response:
`This XML file does not appear to have any style information associated with it. The document tree is shown below.
<WMS_Capabilities xmlns="http://www.opengis.net/wms" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:edal="http://reading-escience-centre.github.io/edal-java/wms" version="1.3.0" updateSequence="2021-11-05T14:22:30.231Z" xsi:schemaLocation="http://www.opengis.net/wms http://schemas.opengis.net/wms/1.3.0/capabilities_1_3_0.xsd">
<Service>
<Name>WMS</Name>
<Title>Thredds Server</Title>
<Abstract>Geospatial Scientific Data</Abstract>
<KeywordList>
<Keyword>Weather and Climate</Keyword>
<Keyword> Food Security and Agriculture</Keyword>
<Keyword> Land Use Land Cover</Keyword>
<Keyword> Water and Water Related Disasters</Keyword>
</KeywordList>
<OnlineResource xlink:type="simple" xlink:href="http://10.5.5.85:8080/thredds/wms/Agg/ucsb-chirps_global_0.05deg_daily.nc4"/>
<ContactInformation>
<ContactPersonPrimary>
<ContactPerson>Francisco Delgado</ContactPerson>
<ContactOrganization>SERVIR SCO</ContactOrganization>
</ContactPersonPrimary>
<ContactVoiceTelephone/>
<ContactElectronicMailAddress>francisco.delgadoolivares@nasa.gov</ContactElectronicMailAddress>
</ContactInformation>
<Fees>none</Fees>
<AccessConstraints>none</AccessConstraints>
<LayerLimit>1</LayerLimit>
<MaxWidth>4096</MaxWidth>
<MaxHeight>4096</MaxHeight>
</Service>
<Capability>
<Request>
<GetCapabilities>
<Format>text/xml</Format>
<DCPType>
<HTTP>
<Get>
<OnlineResource xlink:type="simple" xlink:href="http://10.5.5.85:8080/thredds/wms/Agg/ucsb-chirps_global_0.05deg_daily.nc4"/>
</Get>
</HTTP>
</DCPType>
</GetCapabilities>
<GetMap>
<Format>image/png</Format>
<Format>image/png;mode=32bit</Format>
<Format>image/gif</Format>
<Format>image/jpeg</Format>
<Format>application/vnd.google-earth.kmz</Format>
<DCPType>
<HTTP>
<Get>
<OnlineResource xlink:type="simple" xlink:href="http://10.5.5.85:8080/thredds/wms/Agg/ucsb-chirps_global_0.05deg_daily.nc4"/>
</Get>
</HTTP>
</DCPType>
</GetMap>
<GetFeatureInfo>
<Format>text/plain</Format>
<Format>text/xml</Format>
<Format>text/html</Format>
<DCPType>
<HTTP>
<Get>
<OnlineResource xlink:type="simple" xlink:href="http://10.5.5.85:8080/thredds/wms/Agg/ucsb-chirps_global_0.05deg_daily.nc4"/>
</Get>
</HTTP>
</DCPType>
</GetFeatureInfo>
</Request>
<Exception>
<Format>XML</Format>
</Exception>
<edal:ExtendedCapabilities>
<edal:CapabilitiesType>ncWMS2</edal:CapabilitiesType>
<edal:ExtendedRequest>
<edal:Request>GetMap</edal:Request>
<edal:UrlParameter>
<edal:ParameterName>COLORSCALERANGE</edal:ParameterName>
<edal:ParameterDescription>Of the form min,max this is the scale range used for plotting the data.</edal:ParameterDescription>
</edal:UrlParameter>
<edal:UrlParameter>
<edal:ParameterName>NUMCOLORBANDS</edal:ParameterName>
<edal:ParameterDescription>The number of discrete colours to plot the data. Must be between 2 and 250</edal:ParameterDescription>
</edal:UrlParameter>
<edal:UrlParameter>
<edal:ParameterName>ABOVEMAXCOLOR</edal:ParameterName>
<edal:ParameterDescription>The colour to plot values which are above the maximum end of the scale range. Colours are as defined above, with the addition of "extend", which will use the maximum value of the palette.</edal:ParameterDescription>
</edal:UrlParameter>
<edal:UrlParameter>
<edal:ParameterName>BELOWMINCOLOR</edal:ParameterName>
<edal:ParameterDescription>The colour to plot values which are below the minimum end of the scale range. Colours are as defined above, with the addition of "extend", which will use the minimum value of the palette.</edal:ParameterDescription>
</edal:UrlParameter>
<edal:UrlParameter>
<edal:ParameterName>LOGSCALE</edal:ParameterName>
<edal:ParameterDescription> "true" or "false" - whether to plot data with a logarithmic scale</edal:ParameterDescription>
</edal:UrlParameter>
<edal:UrlParameter>
<edal:ParameterName>TARGETTIME</edal:ParameterName>
<edal:ParameterDescription>For in-situ data, all points which fall within the time range (specified in the TIME parameter) will be plotted. In the case that an in-situ point has multiple time readings within that range, the colour used to plot them will depend on the time value which is closest to this given value</edal:ParameterDescription>
</edal:UrlParameter>
<edal:UrlParameter>
<edal:ParameterName>TARGETELEVATION</edal:ParameterName>
<edal:ParameterDescription>For in-situ data, all points which fall within the elevation range (specified in the ELEVATION parameter) will be plotted. In the case that an in-situ point has multiple elevation readings within that range, the colour used to plot them will depend on the elevation value which is closest to this given value</edal:ParameterDescription>
</edal:UrlParameter>
<edal:UrlParameter>
<edal:ParameterName>OPACITY</edal:ParameterName>
<edal:ParameterDescription>The percentage opacity of the final output image</edal:ParameterDescription>
</edal:UrlParameter>
<edal:UrlParameter>
<edal:ParameterName>ANIMATION</edal:ParameterName>
<edal:ParameterDescription>"true" or "false" - whether to generate an animation. This also needs the TIME to be of the formstarttime/endtime, and currently is only implemented for features with a discrete time axis.</edal:ParameterDescription>
</edal:UrlParameter>
</edal:ExtendedRequest>
<edal:ExtendedRequest>
<edal:Request>GetTimeseries</edal:Request>
<edal:RequestDescription>This produces either a timeseries graph or, if downloading is enabled, a CSV file containing the data. The URL parameters are identical to those of a GetFeatureInfo request. The TIME parameter should specify a range of times in the form starttime/endtime, and the supported formats are: image/png,image/jpg,image/jpeg,text/csv,text/json,application/prs.coverage+json,application/prs.coverage json</edal:RequestDescription>
</edal:ExtendedRequest>
<edal:ExtendedRequest>
<edal:Request>GetVerticalProfile</edal:Request>
<edal:RequestDescription>This produces either a vertical profile graph or, if downloading is enabled, a CSV file containing the data. The URL parameters are identical to those of a GetFeatureInfo request. The ELEVATION parameter should specify a range of elevations in the form startelevation/endelevation, and the supported formats are: image/png,image/jpg,image/jpeg,text/csv,text/json,application/prs.coverage+json,application/prs.coverage json</edal:RequestDescription>
</edal:ExtendedRequest>
<edal:ExtendedRequest>
<edal:Request>GetTransect</edal:Request>
<edal:RequestDescription>This produces a graph of data values along an arbitrary path. Additionally if there is vertical information present in the dataset, it will produce a vertical section along the same path. It accepts the same URL parameters as a GetMap requestion, with the additional mandatory parameter LINESTRING</edal:RequestDescription>
<edal:UrlParameter>
<edal:ParameterName>LINESTRING</edal:ParameterName>
<edal:ParameterDescription>The points which define the path of the transect to plot. Of the form x1 y1,x2 y2,x3 y3...</edal:ParameterDescription>
</edal:UrlParameter>
</edal:ExtendedRequest>
<edal:ExtendedRequest>
<edal:Request>GetMetadata</edal:Request>
<edal:RequestDescription>Fetches small pieces of metadata. Many of these are also present in this capabilities document, but GetMetadata provides a more convenient method of accessing such data. GetMetadata always returns data in the JSON format</edal:RequestDescription>
<edal:UrlParameter>
<edal:ParameterName>ITEM</edal:ParameterName>
<edal:ParameterDescription>This specifies the metadata to return. This can take the values: menu: Returns a tree representation of the available WMS layers, with IDs. Takes the optional parameter DATASET to return the same tree for a single dataset layerDetails: Returns a set of details needed to plot a given layer. This includes such data as units, layer bounding box, configured scale range, etc. Takes the parameters LAYERNAME and TIME. The TIME parameter is optional, and if it is specified then the nearest available time is returned as part of the layer's details. minmax: Calculates the range of values in the given area. Takes the same parameters as a GetMap request. timesteps: Returns the available times for a given day. Takes the parameters LAYERNAME and DAY (yyyy-mm-dd) animationTimesteps: Returns a list of time strings at different temporal resolutions for a given time range. This is used to present to the user different frequencies for the generation of an animation. Takes the parameters LAYERNAME, START, and END</edal:ParameterDescription>
</edal:UrlParameter>
</edal:ExtendedRequest>
<edal:ExtendedRequest>
<edal:Request>GetLegendGraphic</edal:Request>
<edal:RequestDescription>The GetLegendGraphic request generates an image which can be used as a legend. There are two main options: Generating just a colourbar, and generating a full legend.</edal:RequestDescription>
<edal:UrlParameter>
<edal:ParameterName>COLORBARONLY</edal:ParameterName>
<edal:ParameterDescription>"true" or "false". Whether to generate a full legend or just the colour bar. If it's "true", the following URL parameters are required: PALETTE: The name of the palette to use. If missing, set to "default" NUMCOLORBANDS: The number of colour bands to use. If missing, set to 250 VERTICAL: Whether to very colours vertically. If missing, defaults to true WIDTH: The width of the image to generate. If missing, defaults to 50 HEIGHT: The height of the image to generate. If missing, defaults to 200 For a full legend, the additional parameters LAYERS and either STYLES, SLD, or SLD_BODY must be supplied. This is because a single WMS layer may depend on an arbitrary number of sub-layers, depending on the style it is plotted in. In addition to these parameters, the optional parameters controlling the style may be supplied (these are the same as documented in the GetMap request). Note that for full legends, the supplied width and height are NOT the final height of the image, but rather the width and height of each individual coloured plot area (i.e. the 1d/2d colourbar)</edal:ParameterDescription>
</edal:UrlParameter>
</edal:ExtendedRequest>
</edal:ExtendedCapabilities>
<Layer>
<Title>Thredds Server</Title>
<CRS>EPSG:4326</CRS>
<CRS>CRS:84</CRS>
<CRS>EPSG:41001</CRS>
<CRS>EPSG:27700</CRS>
<CRS>EPSG:3408</CRS>
<CRS>EPSG:3409</CRS>
<CRS>EPSG:3857</CRS>
<CRS>EPSG:5041</CRS>
<CRS>EPSG:5042</CRS>
<CRS>EPSG:32661</CRS>
<CRS>EPSG:32761</CRS>
<Layer>
<Title>No dataset title found.</Title>
<Layer queryable="1">
<Name>precipitation_amount</Name>
<Title>precipitation_amount</Title>
<Abstract>precipitation_amount</Abstract>
<EX_GeographicBoundingBox>
<westBoundLongitude>-180.0</westBoundLongitude>
<eastBoundLongitude>180.0</eastBoundLongitude>
<southBoundLatitude>-50.00000149011612</southBoundLatitude>
<northBoundLatitude>50.0</northBoundLatitude>
</EX_GeographicBoundingBox>
<BoundingBox CRS="CRS:84" minx="-180.0" maxx="180.0" miny="-50.00000149011612" maxy="50.0"/>
<Dimension name="time" units="unknown" multipleValues="true" current="true" default="2021-09-30T00:00:00.000Z"> 1981-01-01T00:00:00.000Z/2021-09-30T00:00:00.000Z/P1D </Dimension>
<Style>
<Name>default-scalar/default</Name>
<Title>default-scalar/default</Title>
<Abstract>default-scalar style, using the default palette. Available palettes can be found in the response to http://10.5.5.85:8080/thredds/wms/Agg/ucsb-chirps_global_0.05deg_daily.nc4?request=GetMetadata&item=layerDetails&layerName=precipitation_amount </Abstract>
<LegendURL width="110" height="264">
<Format>image/png</Format>
<OnlineResource xlink:type="simple" xlink:href="http://10.5.5.85:8080/thredds/wms/Agg/ucsb-chirps_global_0.05deg_daily.nc4?REQUEST=GetLegendGraphic&PALETTE=default&LAYERS=precipitation_amount&STYLES=default-scalar/default"/>
</LegendURL>
</Style>
<Style>
<Name>colored_contours/default</Name>
<Title>colored_contours/default</Title>
<Abstract>colored_contours style, using the default palette. Available palettes can be found in the response to http://10.5.5.85:8080/thredds/wms/Agg/ucsb-chirps_global_0.05deg_daily.nc4?request=GetMetadata&item=layerDetails&layerName=precipitation_amount </Abstract>
<LegendURL width="110" height="264">
<Format>image/png</Format>
<OnlineResource xlink:type="simple" xlink:href="http://10.5.5.85:8080/thredds/wms/Agg/ucsb-chirps_global_0.05deg_daily.nc4?REQUEST=GetLegendGraphic&PALETTE=default&LAYERS=precipitation_amount&STYLES=colored_contours/default"/>
</LegendURL>
</Style>
<Style>
<Name>contours</Name>
<Title>contours</Title>
<Abstract>contours style</Abstract>
</Style>
<Style>
<Name>raster/default</Name>
<Title>raster/default</Title>
<Abstract>raster style, using the default palette. Available palettes can be found in the response to http://10.5.5.85:8080/thredds/wms/Agg/ucsb-chirps_global_0.05deg_daily.nc4?request=GetMetadata&item=layerDetails&layerName=precipitation_amount </Abstract>
<LegendURL width="110" height="264">
<Format>image/png</Format>
<OnlineResource xlink:type="simple" xlink:href="http://10.5.5.85:8080/thredds/wms/Agg/ucsb-chirps_global_0.05deg_daily.nc4?REQUEST=GetLegendGraphic&PALETTE=default&LAYERS=precipitation_amount&STYLES=raster/default"/>
</LegendURL>
</Style>
</Layer>
</Layer>
</Layer>
</Capability>
</WMS_Capabilities>`

Also I have tried adding a palette to the default custom palette directory (i had to create it, it was not there) and nothing changed.

**Comment by lesserwhirls** [comment]
So the `precipitation_amount` layer has three styles associated with it:
- `<Name>default-scalar/default</Name>`
- `<Name>colored_contours/default</Name>`
- `<Name>raster/default</Name>`

These are plotting styles and not palettes. If you look in the `<Abstract>` element of each style, you will see a link that will give you a response containing the palettes available (for example: http://10.5.5.85:8080/thredds/wms/Agg/ucsb-chirps_global_0.05deg_daily.nc4?request=GetMetadata&item=layerDetails&layerName=precipitation_amount).

**Comment by lgilliland** [comment]
I see this is closed, but the [v5.4 docs](https://docs.unidata.ucar.edu/tds/5.4/userguide/tds_config_ref.html#wms-service) still show WEB-INF/palettes and how to configure that in the threddsConfig.xml.  Please advise!

**Comment by tdrwenski** [comment]
The "WEB-INF/palettes" is indeed incorrect info, sorry for the confusion. We updated the docs [here](https://github.com/Unidata/tds/pull/342/files), but it looks like we haven't deployed the new docs on the website yet. Let us know if you need further clarification on how to configure the palettes.
